strawberryfields/engine.py

```diff
@@ -127,870 +78,173 @@
             modes (Sequence[int]): Modes to be returned in the state object. If None, returns all modes.
-            apply_history (bool): If True, all applied operations from the previous calls to
-                eng.run are reapplied to the backend state, before applying recently
-                queued quantum operations.
+            compile (bool): if True, compile the Program instances before sending them to the backend
         """
-        if apply_history:
-            self._retain_queue()
-            self.cmd_applied.clear()
-
-        if backend is None:
-            # keep the current backend
-            if self.backend is None:
-                # if backend does not exist, raise error
-                raise ValueError(
-                    "Please provide a simulation backend.") from None
-        elif isinstance(backend, str):
-            # if backend is specified via a string and the engine already has that type of backend
-            # loaded, then we should just use the existing backend
-            # pylint: disable=protected-access
-            if self.backend is not None and self.backend._short_name == backend:
-                pass
-            else:
-                # initialize a backend
-                self.cmd_applied.clear()
-                self.backend = load_backend(backend)
-                self.backend.begin_circuit(
-                    num_subsystems=self.init_num_subsystems, hbar=self.hbar, **kwargs)
-        else:
-            if isinstance(backend, BaseBackend):
-                self.cmd_applied.clear()
-                self.backend = backend
-            else:
-                raise ValueError(
-                    "Please provide a valid Strawberry Fields backend.") from None
-
-        # todo unsuccessful run due to exceptions should ideally result in keeping
-        # the command queue as is but bringing the backend back to the last checkpoint
+        if not isinstance(program, Sequence):
+            program = [program]
+
+        if not self.run_progs:
+            # initialize the backend
+            self.backend.begin_circuit(num_subsystems=program[0].init_num_subsystems, hbar=self.hbar, **kwargs)
+
+        # TODO unsuccessful run due to exceptions should ideally have no effect on the backend state (state checkpoints?)
         try:
-            temp = self._run_command_list(self.cmd_queue, **kwargs)
+            prev = None  # previous program segment
+            for p in program:
+                if self.run_progs:
+                    prev = self.run_progs[-1]
+
+                if prev is not None and not p.can_follow(prev):
+                    raise RuntimeError("Register mismatch after Program {}, '{}'.".format(len(self.run_progs)-1, prev.name))
+
+                # if the program hasn't been compiled for this backend, do it now
+                if compile and p.backend != self.backend._short_name:
+                    p = p.compile(self.backend._short_name)
```

Comment:
What happens if `compile=False`, but the program isn't compiled properly for the target backend? I guess this is something we will need to think about for the new backend API

Reply:
Yes. This was to make some tests work, since compilation can change the relative order of gates that commute.